### ISSUE TITLE
LBaaSv2 / Octavia: Improve LB objects' checks

### ISSUE THREAD

**Issue by kayrus** [terraform-providers/terraform-provider-openstack#546]
Current LBaaSv2 behavior doesn't check for pending LB status, when status check is called from its objects. And if LB is in error state, Terraform will keep trying to update LB objects till reaching the timeout.

Here is an example for terraform-openstack-provider 1.12.0 LB status, when there was an l7policy created manually after the terraform and it is in ERROR state:

```json
{
  "statuses": {
    "loadbalancer": {
      "name": "loadbalancer_1",
      "provisioning_status": "ERROR",
      "listeners": [
        {
          "name": "listener_1",
          "provisioning_status": "ACTIVE",
          "pools": [
            {
              "name": "pool_1",
              "provisioning_status": "ACTIVE",
              "healthmonitor": {},
              "members": [],
              "id": "dc48ef18-ebd9-4c67-81a3-977630fb15e3",
              "operating_status": "ONLINE"
            }
          ],
          "l7policies": [
            {
              "name": "test",
              "provisioning_status": "ERROR",
              "rules": [
                {
                  "type": "PATH",
                  "id": "8f782815-1bb5-4010-a68c-fed5433b0bb6",
                  "provisioning_status": "ACTIVE"
                }
              ],
              "action": "REDIRECT_TO_POOL",
              "id": "8b9b92d1-ebcd-4bff-938e-6d0cae139f4c",
              "operating_status": "DEGRADED"
            }
          ],
          "id": "6399c03f-703e-45d8-8b58-67228bf706ef",
          "operating_status": "DEGRADED"
        }
      ],
      "pools": [
        {
          "name": "pool_1",
          "provisioning_status": "ACTIVE",
          "healthmonitor": {},
          "members": [],
          "id": "dc48ef18-ebd9-4c67-81a3-977630fb15e3",
          "operating_status": "ONLINE"
        }
      ],
      "id": "4a9a810a-3794-4e22-aa16-03356554a43d",
      "operating_status": "DEGRADED"
    }
  }
}
```

I'd like to rename `pool_1` into `pool_2`:

```hcl
resource "openstack_lb_loadbalancer_v2" "loadbalancer_1" {
  name = "loadbalancer_1"
  vip_subnet_id = "11deb7f4-b5ea-478e-bd92-8bba5e678b2e"
}

resource "openstack_lb_pool_v2" "pool_1" {
  name            = "pool_2"
  protocol        = "HTTP"
  lb_method    = "ROUND_ROBIN"
  listener_id = "${openstack_lb_listener_v2.listener_1.id}"
}

resource "openstack_lb_listener_v2" "listener_1" {
  name = "listener_1"
  protocol = "HTTP"
  protocol_port = 8080
  loadbalancer_id = "${openstack_lb_loadbalancer_v2.loadbalancer_1.id}"
}
```

and here is what I get from the `terraform apply`:

```sh
$ terraform apply
openstack_lb_loadbalancer_v2.loadbalancer_1: Refreshing state... (ID: 4a9a810a-3794-4e22-aa16-03356554a43d)
openstack_lb_listener_v2.listener_1: Refreshing state... (ID: 6399c03f-703e-45d8-8b58-67228bf706ef)
openstack_lb_pool_v2.pool_1: Refreshing state... (ID: dc48ef18-ebd9-4c67-81a3-977630fb15e3)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ openstack_lb_pool_v2.pool_1
      name: "pool_1" => "pool_2"


Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

openstack_lb_pool_v2.pool_1: Modifying... (ID: dc48ef18-ebd9-4c67-81a3-977630fb15e3)
  name: "pool_1" => "pool_2"
openstack_lb_pool_v2.pool_1: Still modifying... (ID: dc48ef18-ebd9-4c67-81a3-977630fb15e3, 10s elapsed)
... ... ...
openstack_lb_pool_v2.pool_1: Still modifying... (ID: dc48ef18-ebd9-4c67-81a3-977630fb15e3, 10m0s elapsed)

Error: Error applying plan:

1 error(s) occurred:

* openstack_lb_pool_v2.pool_1: 1 error(s) occurred:

* openstack_lb_pool_v2.pool_1: Error waiting for loadbalancer 4a9a810a-3794-4e22-aa16-03356554a43d to become ACTIVE: timeout while waiting for state to become 'ACTIVE' (last state: 'ERROR', timeout: 10m0s)

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

Expected behavior:

* Ignore LB ERROR status and modify the pool according to its real status

OR

* Stop the terraform apply right away:

```sh
$ terraform apply
openstack_lb_loadbalancer_v2.loadbalancer_1: Refreshing state... (ID: 4a9a810a-3794-4e22-aa16-03356554a43d)
openstack_lb_listener_v2.listener_1: Refreshing state... (ID: 6399c03f-703e-45d8-8b58-67228bf706ef)
openstack_lb_pool_v2.pool_1: Refreshing state... (ID: dc48ef18-ebd9-4c67-81a3-977630fb15e3)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ openstack_lb_pool_v2.pool_1
      name: "pool_1" => "pool_2"


Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

openstack_lb_pool_v2.pool_1: Modifying... (ID: dc48ef18-ebd9-4c67-81a3-977630fb15e3)
  name: "pool_1" => "pool_2"

Error: Error applying plan:

1 error(s) occurred:

* openstack_lb_pool_v2.pool_1: 1 error(s) occurred:

* openstack_lb_pool_v2.pool_1: Error waiting for loadbalancer 4a9a810a-3794-4e22-aa16-03356554a43d to become ACTIVE: unexpected state 'ERROR', wanted target 'ACTIVE'. last error: %!s(<nil>)

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

This is a prerequisite for the #503 issue and #527 PR.

**Comment by kayrus** [comment]
> Ignore LB ERROR status and modify the pool according to its real status

In order to achieve this, we need to to add a `loadbalancer_id` attribute into all LB objects schemas. We will poll LB directly and get the status trees. This will make obsolete all the `waitVia*` functions, but new function like `waitForElement(client, lbID, elementID, target, pending, timeout)` will be created. I need to know whether Octavia returns the whole LB status tree as it is done in Neutron lbaas extension. Probably we can create a multiresource function, which will check statuses only by ID and return best match status.

**Comment by jtopjian** [comment]
First, the error you're seeing is absolutely legitimate and needs fixed.

> In order to achieve this, we need to to add a loadbalancer_id attribute into all LB objects schemas.

But I don't like the idea of doing this. 

If we add `loadbalancer_id` to the schema of the LB resources, would it be a required field? If so, that's going to be a breaking change and will need to be postponed until a v2 release (which isn't on the map).

If it's an optional field, that implies that we can self-determine the load balancer ID internally and therefore doesn't need to exist in the schema, right? Until we can prove this is not true, I don't want to modify the schema.

**Comment by kayrus** [comment]
I suggest `computed` field. What is the `schema migration function`? Never heard about this before.

> self-determine the load balancer ID internally

That is what I'd like to do, but only once, in order to avoid calling Openstack API for every object. Ideally, it would be great to have a shared structure, which could be used by any LB object. Didn't look deep into the Terraform architecture. Is it possible?

**Comment by jtopjian** [comment]
> What is the schema migration function? Never heard about this before.

It's something that should be avoided unless absolutely necessary :)

Here's an example: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_instance_migrate.go

It's a function that will run when a user downloads a new version of the plugin. It walks their schema and applies any updates as necessary. But now that I think about it, this won't work since a schema migration is offline only. So pretend I didn't mentioned this.

> That is what I'd like to do, but only once, in order to avoid calling Openstack API for every object. Ideally, it would be great to have a shared structure, which could be used by any LB object. Didn't look deep into the Terraform architecture. Is it possible?

Unfortunately it's not possible. 

Even if we tried adding some struct to the main config struct (defined in `config.go`), we'll now be sharing reads and writes and that opens up so many cans of worms. We would be dealing with race conditions for months to come.

We must treat each resource as being independent, and if that requires multiple API calls, that's the unfortunate situation we're in. This type of problem is something Terraform is not good at handling. Similar to this is how Terraform cannot take advantage of Octavia's Cascade Delete ability since it's in conflict with how Terraform manages a resource graph internally.

The only way out of this situation would be if there were API calls that would return the LB ID for a given resource:


```
GET /listeners/<uuid>/get_lb_id
GET /l7policies/<uuid>/get_lb_id
```

Or something like that.

As bad as it sounds, it's really not _that_ bad. Knowing the LB ID won't be needed during a `Read` so the user won't see traffic during a `plan` or `refresh` - only when there is a change. Which, to be honest, isn't going to happen _that_ much.

**Comment by kayrus** [comment]
Ok then, I propose the following:

During the resource creation, we can automatically detect LB using the parent object: listener, pool, l7policy (the trickiest one, since openstack doesn't return listener ID for the l7policy). This info could be stored into the new **computed** `loadbalancer_id` attribute. It will be used for fallback status checks, if current object doesn't return the `provisioning_status` value.

For existing terraform states (`tfstate`), the same resolve logic will be added into the `read` function if there is no `loadbalancer_id` defined. Additionally this will allow to introduce resource import mentioned in #521.

If there are no objections, how can I proceed? Within one #547 PR? Or:

* #547 first
* then resolve function + fallback logic
* then #524?

**Comment by jtopjian** [comment]
My spidey sense is going off about this computed field. I understand how you intend to implement it, but it doesn't feel right. The only time I've seen something like this used (in any Terraform plugin) is during `Create` when the information will no longer be available.

My understanding is that the intention here is to effectively cache the `loadbalancer_id`. I understand the merit behind reducing API calls, but compare how many API calls are going to be reduced with how many are already happening (look at all of the GET requests currently happening when you run the acceptance tests with `OS_DEBUG=1` turned on).

By implementing this during Create on the L7 resources as well as implementing this in `Read` for existing resources, we're going to reduce the overall API calls by a few percent points. 

With that said, I am not completely opposed to using this in the new L7 resources in their Create and Import functions. For these new resources, at least we don't have to worry about backporting this feature to existing behavior. I'd need to see the final results of these before signing off on it, though.

Overall, though, if the computed field is simply for caching, then I'm failing to understand its benefit in the overall traffic + additional code required to support it, and I apologize for that.

**Comment by jtopjian** [comment]
The part that is making me so unsure about this field is that the information (`loadbalancer_id`) is always retrievable - it's not an ephemeral piece of information that we need to store before it goes away. 

I'm quite sure that it is not possible to outright change a load balancer ID or move a resource (pool, listener, etc) to a new load balancer, but if there was, this field wouldn't reflect that change. I understand the example I'm using is fictional, but I'm trying to illustrate the principle of always retrieving data from a source of truth rather than creating our own source of truth.

**Comment by kayrus** [comment]
Ok, I'll keep the plan as is. `loadbalancer_id` will be resolved dynamically.

* #547 first
* then resolve function + fallback logic
* then #524
